source/MaterialXRender/OiioImageLoader.cpp

```diff
@@ -78,82 +78,59 @@ bool OiioImageLoader::saveImage(const FilePath& filePath,
+                            image->getResourceBuffer());
             }
             imageOutput->close();
         }
     }
     return written;
 }
 
-bool OiioImageLoader::loadImage(const FilePath& filePath,
-                                ImageDesc& imageDesc,
-                                const ImageDescRestrictions* restrictions)
+ImagePtr OiioImageLoader::loadImage(const FilePath& filePath)
 {
-    imageDesc.width = imageDesc.height = imageDesc.channelCount = 0;
-    imageDesc.resourceBuffer = nullptr;
-
     OIIO::ImageInput* imageInput = OIIO::ImageInput::open(filePath);
     if (!imageInput)
     {
-        return false;
+        return nullptr;
     }
 
     OIIO::ImageSpec imageSpec = imageInput->spec();
+    Image::BaseType baseType = Image::BaseType::UINT8;
     switch (imageSpec.format.basetype)
     {
         case OIIO::TypeDesc::UINT8:
         {
-            imageDesc.baseType = ImageDesc::BASETYPE_UINT8;
+            baseType = Image::BaseType::UINT8;
             break;
         }
         case OIIO::TypeDesc::FLOAT:
         {
-            imageDesc.baseType = ImageDesc::BASETYPE_FLOAT;
+            baseType = Image::BaseType::FLOAT;
             break;
         }
         case OIIO::TypeDesc::HALF:
         {
-            // If 16-bit float is not support try loading in 32-bit float.
-            if (restrictions && restrictions->supportedBaseTypes.count(ImageDesc::BASETYPE_HALF) == 0)
-            {
-                imageSpec.set_format(OIIO::TypeDesc::FLOAT);
-                imageDesc.baseType = ImageDesc::BASETYPE_FLOAT;
-            }
-            else
-            {
-                imageDesc.baseType = ImageDesc::BASETYPE_HALF;
-            }
+            baseType = Image::BaseType::HALF;
             break;
         }
         default:
             break;
     };
 
-    bool read = false;
-    // Check if base type is unsupported before trying to load
-    if (!restrictions || restrictions->supportedBaseTypes.count(imageDesc.baseType))
+    ImagePtr image = Image::create(imageSpec.width, imageSpec.height, imageSpec.nchannel, baseType);
+    size_t imageBytes = (size_t) imageSpec.image_bytes();
+    void* imageBuf = malloc(imageBytes);
+    if (imageInput->read_image(imageSpec.format, imageBuf))
     {
-        imageDesc.width = imageSpec.width;
-        imageDesc.height = imageSpec.height;
-        imageDesc.channelCount = imageSpec.nchannels;
-        imageDesc.computeMipCount();
-
-        size_t imageBytes = (size_t)imageSpec.image_bytes();
-        void* imageBuf = malloc(imageBytes);
-        if (imageInput->read_image(imageSpec.format, imageBuf))
-        {
-            imageDesc.resourceBuffer = imageBuf;
-            imageDesc.resourceBufferDeallocator = [](void* buffer)
-            {
-                free(buffer);
-            };
-            read = true;
-        }
+        image->setResourceBuffer(imageBuf);
+    }
+    else
+    {
+        image = nullptr;
     }
-
     imageInput->close();
```

Comment:
Are you leaking the imageInput here?

Reply:
Good catch, thanks Larry.  It looks like we're missing the call to ImageInput::destroy, and similarly the call to ImageOutput::destroy in the function above.

Reply:
Also, the fact that you're using a raw pointer means it will only work with OIIO 1.x. In 2.x it changed to a unique_ptr. The 1.x/2.x agnostic idiom is 

    auto imageInput = OIIO::ImageInput::open(filePath); // works for raw or unique_ptr
    ... all rest the same...
    OIIO::ImageInput::destroy(imageInput); // destroy is superfluous once min is >= 2.0

Reply:
Ah, thanks for the tip!  We’ll plan to update this to the version-agnostic idiom for future compatibility.

Reply:
... or even past compatibility. OIIO 2.0 is already a full year old and 2.1 is about to be in release candidate stage. :-)